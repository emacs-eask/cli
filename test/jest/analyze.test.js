const { TestContext } = require("./helpers");

describe("analyze", () => {
  describe("in ./analyze/dsl", () => {
    const ctx = new TestContext("./test/jest/analyze/dsl");

    it("handles plain text", async () => {
      await ctx.runEask("analyze");
      await ctx.runEask("analyze Eask");
    });

    it("handles json option", async () => {
      const { stderr } = await ctx.runEask("analyze --json");
      await ctx.runEask("analyze Eask --json");

      // try to parse output, errors if not valid
      JSON.parse(stderr);
    });

    it("matches snapshot", async () => {
      const res = await ctx.runEask("analyze");
      const resClean = res.sanitized().raw();
      expect(resClean).toMatchSnapshot();
    });

    it("should report multiple definitions", async () => {
      const { stderr } = await ctx.runEask("analyze");
      // expect this substring
      expect(stderr).toMatch("Multiple definition of `package'");
    });
  });

  describe("in ./empty", () => {
    const ctx = new TestContext("./test/jest/empty");

    it("should error", async () => {
      await expect(ctx.runEask("analyze")).rejects.toThrow();
    });
  });

  describe("in ./analyze/metadata", () => {
    const ctx = new TestContext("./test/jest/analyze/metadata");

    it("handles plain text", async () => {
      await ctx.runEask("analyze");
      await ctx.runEask("analyze Eask");
    });

    it("handles json", async () => {
      await ctx.runEask("analyze --json");
      await ctx.runEask("analyze Eask --json");
    });
  });

  describe("in ./analyze/errors", () => {
    const ctx = new TestContext("./test/jest/analyze/errors");

    // Eask-normal - no errors or warnings
    // Eask-warn   - only warnings
    // Eask-error  - errors and  warnings

    it("should check Eask-normal", async () => {
      await ctx.runEask("analyze Eask-normal");
    });

    it("should check Eask-warn", async () => {
      await ctx.runEask("analyze Eask-warn");
    });

    it.failing("should error on Eask-errors", async () => {
      await expect(ctx.runEask("analyze Eask-error")).rejects.toThrow();
    });

    it.failing("should error when using --strict on Eask-warn", async () => {
      await expect(ctx.runEask("analyze --strict Eask-warn")).rejects.toThrow();
    });

    // sanity check: flag should not change behavior in this case
    // this is because the menaing of --allow-error is "continue to the end"
    it.failing("should error when --allow-error is set", async () => {
      await expect(
        ctx.runEask("analyze Eask-error --allow-error"),
      ).rejects.toThrow();
    });

    it.failing("should check all files when --allow-error is set", async () => {
      await expect(
        ctx.runEask("analyze --allow-error Eask-normal Eask-error"),
      ).rejects.toMatchObject({ stderr: "(Checked 2 files)" });
    });
  });
});
