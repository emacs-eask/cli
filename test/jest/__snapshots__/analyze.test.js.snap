// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`analyze in ./analyze/dsl handles json option 1`] = `
"{
  "warnings": [
    {
      "range": {
        "start": {
          "line": 25,
          "col": 0,
          "pos": 853
        },
        "end": {
          "line": 25,
          "col": 35,
          "pos": 888
        }
      },
      "filename": "/home/naver/git/eask-cli/test/jest/analyze/dsl/Eask",
      "message": "ðŸ’¡ Pkg-file seems to be missing \`check-pkg.el'"
    },
    {
      "range": {
        "start": {
          "line": 20,
          "col": 0,
          "pos": 709
        },
        "end": {
          "line": 20,
          "col": 17,
          "pos": 726
        }
      },
      "filename": "/home/naver/git/eask-cli/test/jest/analyze/dsl/Eask",
      "message": "ðŸ’¡ Warning regarding duplicate license name, GPLv3"
    },
    {
      "range": {
        "start": {
          "line": 17,
          "col": 0,
          "pos": 613
        },
        "end": {
          "line": 17,
          "col": 15,
          "pos": 628
        }
      },
      "filename": "/home/naver/git/eask-cli/test/jest/analyze/dsl/Eask",
      "message": "ðŸ’¡ Warning regarding duplicate author name, name"
    }
  ],
  "errors": [
    {
      "range": {
        "start": {
          "line": 46,
          "col": 0,
          "pos": 1696
        },
        "end": {
          "line": 50,
          "col": 2,
          "pos": 1908
        }
      },
      "filename": "/home/naver/git/eask-cli/test/jest/analyze/dsl/Eask",
      "message": "âœ— Define dependencies with the same name \`f' with different version"
    },
    {
      "range": {
        "start": {
          "line": 46,
          "col": 0,
          "pos": 1696
        },
        "end": {
          "line": 50,
          "col": 2,
          "pos": 1908
        }
      },
      "filename": "/home/naver/git/eask-cli/test/jest/analyze/dsl/Eask",
      "message": "âœ— Define dependencies with the same name \`f'"
    },
    {
      "range": {
        "start": {
          "line": 44,
          "col": 0,
          "pos": 1596
        },
        "end": {
          "line": 44,
          "col": 27,
          "pos": 1623
        }
      },
      "filename": "/home/naver/git/eask-cli/test/jest/analyze/dsl/Eask",
      "message": "âœ— Define dependencies with the same name \`dash' with different version"
    },
    {
      "range": {
        "start": {
          "line": 43,
          "col": 0,
          "pos": 1576
        },
        "end": {
          "line": 43,
          "col": 19,
          "pos": 1595
        }
      },
      "filename": "/home/naver/git/eask-cli/test/jest/analyze/dsl/Eask",
      "message": "âœ— Define dependencies with the same name \`dash'"
    },
    {
      "range": {
        "start": {
          "line": 40,
          "col": 0,
          "pos": 1474
        },
        "end": {
          "line": 40,
          "col": 20,
          "pos": 1494
        }
      },
      "filename": "/home/naver/git/eask-cli/test/jest/analyze/dsl/Eask",
      "message": "âœ— Define dependencies with the same name \`emacs'"
    },
    {
      "range": {
        "start": {
          "line": 37,
          "col": 0,
          "pos": 1304
        },
        "end": {
          "line": 37,
          "col": 15,
          "pos": 1319
        }
      },
      "filename": "/home/naver/git/eask-cli/test/jest/analyze/dsl/Eask",
      "message": "âœ— Unknown package archive \`local'"
    },
    {
      "range": {
        "start": {
          "line": 37,
          "col": 0,
          "pos": 1304
        },
        "end": {
          "line": 37,
          "col": 15,
          "pos": 1319
        }
      },
      "filename": "/home/naver/git/eask-cli/test/jest/analyze/dsl/Eask",
      "message": "âœ— Invalid archive name \`local'"
    },
    {
      "range": {
        "start": {
          "line": 35,
          "col": 0,
          "pos": 1229
        },
        "end": {
          "line": 35,
          "col": 24,
          "pos": 1253
        }
      },
      "filename": "/home/naver/git/eask-cli/test/jest/analyze/dsl/Eask",
      "message": "âœ— Unknown package archive \`magic-archive'"
    },
    {
      "range": {
        "start": {
          "line": 33,
          "col": 0,
          "pos": 1214
        },
        "end": {
          "line": 33,
          "col": 13,
          "pos": 1227
        }
      },
      "filename": "/home/naver/git/eask-cli/test/jest/analyze/dsl/Eask",
      "message": "âœ— Multiple definition of source \`gnu'"
    },
    {
      "range": {
        "start": {
          "line": 30,
          "col": 0,
          "pos": 1073
        },
        "end": {
          "line": 30,
          "col": 61,
          "pos": 1134
        }
      },
      "filename": "/home/naver/git/eask-cli/test/jest/analyze/dsl/Eask",
      "message": "âœ— Run-script with the same key name is not allowed: \`test\`"
    },
    {
      "range": {
        "start": {
          "line": 26,
          "col": 0,
          "pos": 953
        },
        "end": {
          "line": 26,
          "col": 35,
          "pos": 988
        }
      },
      "filename": "/home/naver/git/eask-cli/test/jest/analyze/dsl/Eask",
      "message": "âœ— Multiple definition of \`package-descriptor'"
    },
    {
      "range": {
        "start": {
          "line": 23,
          "col": 0,
          "pos": 822
        },
        "end": {
          "line": 23,
          "col": 29,
          "pos": 851
        }
      },
      "filename": "/home/naver/git/eask-cli/test/jest/analyze/dsl/Eask",
      "message": "âœ— Multiple definition of \`package-file'"
    },
    {
      "range": {
        "start": {
          "line": 14,
          "col": 0,
          "pos": 517
        },
        "end": {
          "line": 14,
          "col": 16,
          "pos": 533
        }
      },
      "filename": "/home/naver/git/eask-cli/test/jest/analyze/dsl/Eask",
      "message": "âœ— Multiple definition of \`keywords'"
    },
    {
      "range": {
        "start": {
          "line": 12,
          "col": 0,
          "pos": 383
        },
        "end": {
          "line": 12,
          "col": 55,
          "pos": 438
        }
      },
      "filename": "/home/naver/git/eask-cli/test/jest/analyze/dsl/Eask",
      "message": "âœ— Multiple definition of \`website-url'"
    },
    {
      "range": {
        "start": {
          "line": 9,
          "col": 0,
          "pos": 290
        },
        "end": {
          "line": 9,
          "col": 18,
          "pos": 308
        }
      },
      "filename": "/home/naver/git/eask-cli/test/jest/analyze/dsl/Eask",
      "message": "âœ— Multiple definition of \`package'"
    }
  ]
}
"
`;

exports[`analyze in ./analyze/dsl matches snapshot 1`] = `
{
  "stderr": "

[36m(Checked 1 file)[0m
",
  "stdout": "~/Eask:9:18 Error: âœ— Multiple definition of \`package'
~/Eask:12:55 Error: âœ— Multiple definition of \`website-url'
~/Eask:14:16 Error: âœ— Multiple definition of \`keywords'
~/Eask:17:15 Warning: ðŸ’¡ Warning regarding duplicate author name, name
~/Eask:20:17 Warning: ðŸ’¡ Warning regarding duplicate license name, GPLv3
~/Eask:23:29 Error: âœ— Multiple definition of \`package-file'
~/Eask:25:35 Warning: ðŸ’¡ Pkg-file seems to be missing \`check-pkg.el'
~/Eask:26:35 Error: âœ— Multiple definition of \`package-descriptor'
~/Eask:30:61 Error: âœ— Run-script with the same key name is not allowed: \`test\`
~/Eask:33:13 Error: âœ— Multiple definition of source \`gnu'
~/Eask:35:24 Error: âœ— Unknown package archive \`magic-archive'
~/Eask:37:15 Error: âœ— Invalid archive name \`local'
~/Eask:37:15 Error: âœ— Unknown package archive \`local'
~/Eask:40:20 Error: âœ— Define dependencies with the same name \`emacs'
~/Eask:43:19 Error: âœ— Define dependencies with the same name \`dash'
~/Eask:44:27 Error: âœ— Define dependencies with the same name \`dash' with different version
~/Eask:50:2 Error: âœ— Define dependencies with the same name \`f'
~/Eask:50:2 Error: âœ— Define dependencies with the same name \`f' with different version
",
}
`;
